dlc-oracle/src/main/scala/org/bitcoins/dlc/oracle/DLCOracle.scala

```diff
@@ -0,0 +1,224 @@
+    require(path.forall(_.hardened),
+            s"Cannot use a BIP32Path with unhardened nodes, got $path")
+    val priv = extPrivateKey.deriveChildPrivKey(path).key
+    val hash =
+      CryptoUtil.sha256(
+        priv.schnorrNonce.bytes ++ CryptoUtil.serializeForHash(label))
+    val tweak = ECPrivateKey(hash.bytes)
+
+    priv.add(tweak)
+  }
+
+  def listEventDbs(): Future[Vector[EventDb]] = eventDAO.findAll()
+
+  def listPendingEventDbs(): Future[Vector[EventDb]] = eventDAO.getPendingEvents
+
+  def listEvents(): Future[Vector[Event]] = {
+    for {
+      rValDbs <- rValueDAO.findAll()
+      eventDbs <- eventDAO.findAll()
+      outcomes <- eventOutcomeDAO.findAll()
+    } yield {
+      val rValDbsByNonce = rValDbs.groupBy(_.nonce)
+      val outcomesByNonce = outcomes.groupBy(_.nonce)
+      eventDbs.map(db =>
+        Event(rValDbsByNonce(db.nonce).head, db, outcomesByNonce(db.nonce)))
+    }
+  }
+
+  def getEvent(nonce: SchnorrNonce): Future[Option[Event]] = {
+    for {
+      rValDbOpt <- rValueDAO.read(nonce)
+      eventDbOpt <- eventDAO.read(nonce)
+      outcomes <- eventOutcomeDAO.findByNonce(nonce)
+    } yield {
+      (rValDbOpt, eventDbOpt) match {
+        case (Some(rValDb), Some(eventDb)) =>
+          Some(Event(rValDb, eventDb, outcomes))
+        case (None, None) | (Some(_), None) | (None, Some(_)) =>
+          None
+      }
+    }
+  }
+
+  def createNewEvent(
+      eventName: String,
+      outcomes: Vector[String]): Future[EventDb] = {
+    for {
+      indexOpt <- rValueDAO.maxKeyIndex
+      index = indexOpt match {
+        case Some(value) => value + 1
+        case None        => 0
+      }
+
+      path = getPath(index)
+      nonce = getKValue(eventName, path).schnorrNonce
+
+      hash =
+        CryptoUtil.sha256(nonce.bytes ++ CryptoUtil.serializeForHash(eventName))
+      commitmentSig = signingKey.schnorrSign(hash.bytes)
+
+      rValueDb = RValueDbHelper(nonce = nonce,
+                                eventName = eventName,
+                                account = rValueAccount,
+                                chainType = chainIndex,
+                                keyIndex = index,
+                                commitmentSignature = commitmentSig)
+
+      eventDb = EventDb(nonce, eventName, outcomes.size, Mock, None)
+      eventOutcomeDbs = outcomes.map { outcome =>
+        val hash = CryptoUtil.sha256(outcome)
+        EventOutcomeDb(nonce, outcome, hash)
+      }
+
+      _ <- rValueDAO.create(rValueDb)
+      eventDb <- eventDAO.create(eventDb)
+      _ <- eventOutcomeDAO.createAll(eventOutcomeDbs)
+    } yield eventDb
+  }
+
+  def signEvent(nonce: SchnorrNonce, outcome: String): Future[EventDb] = {
+    for {
+      rValDbOpt <- rValueDAO.read(nonce)
+      rValDb = rValDbOpt match {
+        case Some(value) => value
+        case None =>
+          throw new RuntimeException(
+            s"Nonce not found from this oracle ${nonce.hex}")
+      }
```

Comment:
nit: should be
```
rValDb <- rValDbOpt match {
  case Some(value) => Future.successful(value)
  case None => Future.failed(new RuntimeException(s"Nonce not found for this oracle ${nonce.hex}")
}
```
same below

Reply:
I think these are functionally the same because it is already in a `flatMap`

Reply:
Implemented anyways

Reply:
functionally the same but throw should be avoided when possible as it is terribly inefficient

Reply:
TIL